urllib3/util/selectors.py

```diff
@@ -535,19 +535,30 @@ def close(self):
             super(KqueueSelector, self).close()
 
 
+if not hasattr(select, 'select'):  # Platform-specific: AppEngine
+    HAS_SELECT = False
+
+
 # Choose the best implementation, roughly:
 # kqueue == epoll > poll > select. Devpoll not supported. (See above)
 # select() also can't accept a FD > FD_SETSIZE (usually around 1024)
-if 'KqueueSelector' in globals():  # Platform-specific: Mac OS and BSD
-    DefaultSelector = KqueueSelector
-elif 'EpollSelector' in globals():  # Platform-specific: Linux
-    DefaultSelector = EpollSelector
-elif 'PollSelector' in globals():  # Platform-specific: Linux
-    DefaultSelector = PollSelector
-elif 'SelectSelector' in globals():  # Platform-specific: Windows
-    DefaultSelector = SelectSelector
-else:  # Platform-specific: AppEngine
-    def no_selector(_):
-        raise ValueError("Platform does not have a selector")
-    DefaultSelector = no_selector
-    HAS_SELECT = False
+def _default_selector(*_):
+    """ This function serves as a first call for DefaultSelector to
+    detect if the select module is being monkey-patched incorrectly
+    by eventlet, greenlet, and preserve proper behavior. """
+    global DefaultSelector
+    if hasattr(select, 'kqueue'):  # Platform-specific: Mac OS
+        selector_type = KqueueSelector
+    elif hasattr(select, 'epoll'):  # Platform-specific: Linux
+        selector_type = EpollSelector
+    elif hasattr(select, 'poll'):  # Platform-specific: Linux
+        selector_type = PollSelector
+    elif hasattr(select, 'select'):  # Platform-specific: Windows
+        selector_type = SelectSelector
+    else:  # Platform-specific: AppEngine
+        raise ValueError('Platform does not have a selector')
+    DefaultSelector = selector_type
```

Comment:
Do we only want to use this function the first time we initialize a selector, or every time? What if a selector gets used before the monkey patch takes effect? What sort of performance costs would we incur by simply leaving DefaultSelector set to execute this function, rather than attaching to a concrete selector class?